File: baekjoon/DFS&BFS/1325_효율적인 해킹.js
const fs = require('fs')
const [nm, ...input] = fs
  .readFileSync('../input.txt')
  .toString()
  .trim()
  .split('\n')
const [n, m] = nm.split(' ').map(Number)
const array = input.map((el) => el.split(' ').map(Number))

function bfs(idx, graph) {
  let cnt = 0 // 해킹할 수 있는 컴퓨터 수
  const queue = [idx]
  // 이미 해킹한 것은 제외해야하기 때문에 방문자 배열 이용
  const visited = Array(n + 1).fill(false)
  visited[idx] = true

  while (queue.length) {
    const tmp = queue.shift()
    for (let i = 0; i < graph[tmp].length; i++) {
      const com = graph[tmp][i]
      if (visited[com] === false) {
        visited[com] = true
        queue.push(com)
        cnt += 1
      }
    }
  }

  return cnt
}

function solution() {
  let computer = Array(n + 1).fill(0) // 해킹한 컴퓨터 인덱스 = 해킹할 수 있는 컴퓨터 수
  const graph = Array.from(Array(n + 1), () => [])

  for (const [x, y] of array) {
    graph[y].push(x)
  }

  for (let idx = 1; idx <= n; idx++) {
    const cnt = bfs(idx, graph)
    // idx번째 컴퓨터가 해킹 할 수 있는 수
    computer[idx] = cnt
  }

  const max = Math.max(...computer)
  const answer = []
  // 최대로 해킹할 수 있는 컴퓨터를 찾아 컴퓨터 번호 저장
  computer.forEach((com, idx) => (com === max ? answer.push(idx) : ''))

  console.log(...answer)
}

solution()

File: baekjoon/DFS&BFS/1325_효율적인 해킹.py
from collections import deque
import sys
input = sys.stdin.readline
n, m = map(int, input().split())
array = [list(map(int, input().split())) for _ in range(m)]


def bfs(idx, graph):
    cnt = 0
    queue = deque()
    queue.append(idx)
    visited = [False] * (n+1)
    visited[idx] = True

    while queue:
        tmp = queue.popleft()
        for i in graph[tmp]:
            if visited[i] == False:
                visited[i] = True
                queue.append(i)
                cnt += 1

    return cnt


def solution():
    computer = []
    compare = 0
    graph = [[] for _ in range(n+1)]

    for [x, y] in array:
        graph[y].append(x)

    for idx in range(1, n+1):
        cnt = bfs(idx, graph)
        computer.append([idx, cnt])
        compare = max(compare, cnt)

    answer = []
    for [com, cnt] in computer:
        if cnt == compare:
            answer.append(com)
    answer.sort()

    print(*answer)


solution()
